Update pyproject (python3.12 support)

Sicelukwanda opened the pull request:
This PR updates project metadata, dependencies, and build system configuration to support [pixi](https://pixi.sh/latest/reference/pixi_manifest/#pixi-build). Updated the target Python version for `ruff` to 3.12.

blooop committed:
set setuptools to >69.0.0

blooop committed:
[bugfix] use a default factor for mutable values

Sicelukwanda committed:
update target version to Python 3.12 in pyproject.toml

Sicelukwanda committed:
add python version requirements and classifiers in setup.py

Sicelukwanda committed:
add minimal build-system section to pyproject.toml

Sicelukwanda committed:
add minimal pixi project configuration to pyproject.toml

Sicelukwanda committed:
update project dependencies in pyproject.toml

Sicelukwanda committed:
Add project metadata tables for pixi in toml

Sicelukwanda committed:
Loosen python_requires in setup.py to also accept 3.12.*

Sicelukwanda committed:
Loosen python_requires in pyproject.toml to also accept 3.12.*

Sicelukwanda committed:
Remove licence field

Sicelukwanda committed:
Merge branch 'main' into feature/setuptools_default_factory

blooop commented on pyproject.toml:
```diff
@@ -1,5 +1,33 @@
+[project]
+name = "gripx"
+version = "0.0.9.dev"
```
I don't know how this works/conficts with the existion version defined in `version.py`  I know you can set up pyproject.toml to have a dynamic version as defined in another file.

Sicelukwanda replied:
Fixed :+1:

blooop commented on pyproject.toml:
```diff
@@ -64,4 +92,25 @@ ignored-modules = ["rclpy",
+name = "gripx"
+channels = ["conda-forge"]
+platforms = ["linux-64"]
+preview = ["pixi-build"]
```
what is preview for?

Sicelukwanda replied:
preview was needed to let pixi use direct dependencies, i.e., `gripx = {git = "https://github.com/Sicelukwanda/grip.git", rev = "feature/setuptools_default_factory"}`

blooop commented on pyproject.toml:
```diff
@@ -64,4 +92,25 @@ ignored-modules = ["rclpy",
+
+[tool.pixi.package]          # ← note the prefix
+name = "gripx"
+version = "0.0.9.dev"
```
another duplicate of version

blooop commented on pyproject.toml:
```diff
@@ -64,4 +92,25 @@ ignored-modules = ["rclpy",
+  "https://prefix.dev/conda-forge",
+]
+
+[tool.pixi.package.host-dependencies]
```
is this really required?

Sicelukwanda committed:
Remove unnecessary pixi tables in pyproject.toml file

Sicelukwanda commented:
@blooop, I had added a minimal pixi manifest within the toml file to tell pixi (when used within gym-kr1) how it should build gripx. Without these it was defaulting to the version of gripx on pypi and failing because the pypi version is tied to python 3.10. 

I've now removed all the pixi stuff because we are no longer trying to work off a specific branch (feature not quite supported by pixi).

blooop commented on grip/environments/domain_randomiser.py:
```diff
@@ -2,7 +2,7 @@
 import pybullet as p
 import numpy as np
-from typing import Union, Tuple, List
+from typing import Union, Tuple, List, Callable
```
don't need Callable

Sicelukwanda committed:
Remove unused Callable import from domain_randomiser.py
